File: benchmarks/deno/st-mergesort.ts
import { readFileSync, writeFileSync } from 'node:fs';

function merge(arr: Int32Array, left: number, mid: number, right: number): void {
    const n1 = mid - left + 1;
    const n2 = right - mid;

    // Create temporary arrays
    const L = new Int32Array(n1);
    const R = new Int32Array(n2);

    // Copy data to temporary arrays using memcpy-like approach
    L.set(arr.slice(left, left + n1));
    R.set(arr.slice(mid + 1, mid + 1 + n2));

    // Merge the temporary arrays back
    let i = 0, j = 0, k = left;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    // Copy remaining elements of L[]
    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    // Copy remaining elements of R[]
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

function mergeSort(arr: Int32Array, left: number, right: number): void {
    if (left < right) {
        const mid = Math.floor(left + (right - left) / 2);
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

// Main function
if (Deno.args.length !== 3) {
    console.error('Usage: deno run st-mergesort.ts <input_file> <num_integers> <output_file>');
    Deno.exit(1);
}

const inputFile = Deno.args[0];
const numIntegers = parseInt(Deno.args[1]);
const outputFile = Deno.args[2];

try {
    // Read input file
    const buffer = await Deno.readFile(inputFile);
    if (buffer.length < numIntegers * 4) {
        console.error('Error: Input file is too small');
        Deno.exit(1);
    }

    // Create array directly from buffer, similar to C's fread
    const arr = new Int32Array(buffer.buffer, buffer.byteOffset, numIntegers);

    // Perform merge sort
    mergeSort(arr, 0, numIntegers - 1);

    // Write output file directly, similar to C's fwrite
    await Deno.writeFile(outputFile, new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength));
} catch (error) {
    console.error('Error:', error.message);
    Deno.exit(1);
}

File: benchmarks/graalvm/StMergeSort.java
import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.IntBuffer;

public class StMergeSort {
    public static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        // Create temporary arrays
        int[] L = new int[n1];
        int[] R = new int[n2];

        // Copy data to temporary arrays
        System.arraycopy(arr, left, L, 0, n1);
        System.arraycopy(arr, mid + 1, R, 0, n2);

        // Merge the temporary arrays back
        int i = 0, j = 0, k = left;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        // Copy remaining elements of L[]
        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }

        // Copy remaining elements of R[]
        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = left + (right - left) / 2;
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);
            merge(arr, left, mid, right);
        }
    }

    public static void main(String[] args) {
        if (args.length != 3) {
            System.err.println("Usage: java StMergeSort <input_file> <num_integers> <output_file>");
            System.exit(1);
        }

        String inputFile = args[0];
        int numIntegers = Integer.parseInt(args[1]);
        String outputFile = args[2];

        try {
            // Read input file
            FileInputStream fis = new FileInputStream(inputFile);
            byte[] bytes = new byte[numIntegers * 4];
            fis.read(bytes);
            fis.close();

            // Convert bytes to integers
            ByteBuffer bb = ByteBuffer.wrap(bytes);
            bb.order(ByteOrder.LITTLE_ENDIAN);
            IntBuffer ib = bb.asIntBuffer();
            int[] arr = new int[numIntegers];
            ib.get(arr);

            // Perform merge sort
            mergeSort(arr, 0, numIntegers - 1);

            // Write output file
            FileOutputStream fos = new FileOutputStream(outputFile);
            ByteBuffer outBuffer = ByteBuffer.allocate(numIntegers * 4);
            outBuffer.order(ByteOrder.LITTLE_ENDIAN);
            outBuffer.asIntBuffer().put(arr);
            fos.write(outBuffer.array());
            fos.close();

        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
}
